### PR TITLE
Add a evaluateForEffectsWithPriorEffects to Realm

### PR DESCRIPTION
**Pull request by trueadm**
Release notes: none

This should fix https://github.com/facebook/prepack/issues/1803 and extends from the unsuccessful attempts made in https://github.com/facebook/prepack/pull/1809 to fix this.

When we apply the `priorEffects` inside of the `evaluateForEffects` on TryStatement, we have to ensure that the `canBeApplied` flag is reset (as the bindings on the Realm get restored by `evaluateForEffects`).

### CODE REVIEW

**Comment by gaearon** [comment]
Why is it safe to reset the flag in this particular case?

`withEffectsAppliedInGlobalEnv` also called `restoreBindings` and `restoreProperties`. We're not doing it here. Why?

**Comment by gaearon** [comment]
Also why not use `withEffectsAppliedInGlobalEnv` directly here?

**Comment by trueadm** [comment]
That's strange, you're seeing my old commit not my new. One sec.

**Comment by trueadm** [comment]
@gaearon The reason for not using `withEffectsAppliedInGlobalEnv` directly was because it does a bunch of logic that would conflict with the logic in `TryStatement` – specifically it applies the effects passed in, when we don't want that as the logic in `TryStatement` is to apply a list of `priorEffects`.

**Comment by gaearon** [comment]
Does the test pass for you? When I run this as `input.js` I still get the invariant (but in a new place now).

**Comment by trueadm** [comment]
I've tidied up the solution and added comments.

File: src/evaluators/TryStatement.js [hunk 1]
```diff
@@ -22,6 +22,30 @@ import { Functions, Join } from "../singletons.js";
 import { Value } from "../values/index.js";
 import type { BabelNodeTryStatement } from "babel-types";
 import invariant from "../invariant.js";
+import { generate } from "rxjs/observable/generate";
+
+// This function gets the evaluated effects with a collection of
+// prior nested affects applied (and their canBeApplied flag reset)
+// We can safely do this as we're wrapped the effects in evaluated
```
Comment (gaearon) [inline src/evaluators/TryStatement.js]: we're -> we've?

File: src/evaluators/TryStatement.js [hunk 2]
```diff
@@ -22,6 +22,30 @@ import { Functions, Join } from "../singletons.js";
...
+  realm: Realm,
+  priorEffects: Array<Effects>,
+  f: () => AbruptCompletion | Value,
+  generatorName: string
+): Effects {
+  return realm.evaluateForEffects(
+    () => {
+      for (let priorEffect of priorEffects) realm.applyEffects(priorEffect);
+      let v = f();
+      for (let priorEffect of priorEffects) priorEffect.canBeApplied = true;
```
Comment (gaearon) [inline src/evaluators/TryStatement.js]: Maybe add an invariant here that it was `false`?

File: src/evaluators/TryStatement.js [hunk 3]
```diff
@@ -23,6 +23,32 @@ import { Value } from "../values/index.js";
...
+function evaluateForEffectsWithPriorEffects(
+  realm: Realm,
+  priorEffects: Array<Effects>,
+  f: () => AbruptCompletion | Value,
+  generatorName: string
+): Effects {
+  return realm.evaluateForEffects(
+    () => {
+      for (let priorEffect of priorEffects) realm.applyEffects(priorEffect);
+      let v = f();
```
Comment (gaearon) [inline src/evaluators/TryStatement.js]: What if `f` throws? Should restoration happen inside a `finally` block, and `f()` be inside of a `try` block?

Comment (gaearon) [reply src/evaluators/TryStatement.js]: (I'm looking at `withEffectsAppliedInGlobalEnv` which does it this way)

File: src/evaluators/TryStatement.js [hunk 4]
```diff
@@ -174,14 +201,14 @@ export default function(ast: BabelNodeTryStatement, strictCode: boolean, env: Le
...
+        realm,
+        priorEffects,
         () => {
-          for (let priorEffect of priorEffects) realm.applyEffects(priorEffect);
           invariant(ast.finalizer);
           return env.evaluateCompletionDeref(ast.finalizer, strictCode);
         },
-        undefined,
-        "composeNestedThrowEffectsWithHandler"
+        "composeNestedThrowEffectsWithHandler/2"
```
Comment (gaearon) [inline src/evaluators/TryStatement.js]: Let's fix this to be `composeNestedEffectsWithFinalizer/2`. Looks like bad original copypasta.

**Comment by gaearon** [comment]
@NTillmann This is as much as I can reduce, wdyt? https://github.com/facebook/prepack/pull/1811/commits/9abfd06285fd445fd6af3752abfca2ecf3ca778d

File: src/evaluators/TryStatement.js [hunk 5]
```diff
@@ -23,6 +23,33 @@ import { Value } from "../values/index.js";
 import type { BabelNodeTryStatement } from "babel-types";
 import invariant from "../invariant.js";
 
+// This function gets the evaluated effects with a collection of
+// prior nested affects applied (and their canBeApplied flag reset)
+// We can safely do this as we've wrapped the effects in evaluated
+// effects, meaning all the effects applied to Realm get restored
+function evaluateForEffectsWithPriorEffects(
```
Comment (hermanventer) [inline src/evaluators/TryStatement.js]: This seems like something that belongs in Realm.

**Pull request closed without merging**